src/System.Buffers.Primitives/System/Internal/OwnedArray.cs

```diff
@@ -68,20 +68,21 @@ public override void Retain()
             Interlocked.Increment(ref _referenceCount);
         }
 
-        public override void Release()
+        public override bool Release()
         {
             if (!IsRetained) BufferPrimitivesThrowHelper.ThrowInvalidOperationException();
             if (Interlocked.Decrement(ref _referenceCount) == 0)
             {
                 OnNoReferences();
             }
+            return IsRetained;
```

Comment:
This is racy and has the potential for conflict. The interlocked operation above could bring the ref count to a value above 0 (i.e. this code does not do the disposed) but by the time this line of code runs, someone else did bring it down to 0. Then, both callers get a result of true. If the one that didn't end up releasing it makes some decision based on the belief that they did (via the return value), that could cause an issue.